src/Conversion/ONNXToTOSA/DialectBuilder.cpp

```diff
@@ -113,6 +113,30 @@ Value TosaBuilder::slice(Value &inputConst, llvm::ArrayRef<int64_t> size,
   return newSliceInput;
 }
 
+Value TosaBuilder::reshape(mlir::Value &value, llvm::ArrayRef<int64_t> shape) {
+  auto shapeAttr = rewriter().getDenseI64ArrayAttr(shape);
+  auto valueType = value.getType().cast<ShapedType>();
+  Type newValueType = RankedTensorType::get(
+      llvm::SmallVector<int64_t, 4>(shape.size(), ShapedType::kDynamic),
+      valueType.getElementType());
+  return tosa::CreateOpAndInfer<mlir::tosa::ReshapeOp>(
+      rewriter(), loc(), newValueType, value, shapeAttr);
+}
+
+Value TosaBuilder::mul(mlir::Value &lhs, mlir::Value &rhs, int32_t shift) {
+  auto lhsType = lhs.getType().cast<ShapedType>();
+  auto rhsType = rhs.getType().cast<ShapedType>();
+  assert(lhsType.getElementType() == rhsType.getElementType() &&
+         "lhs and rhs have different element types");
+  auto outputRank = lhsType.getRank() > rhsType.getRank() ? lhsType.getRank()
```

Comment:
Thank you for the comment. I added that functionality to the `TosaBuilder`. I actually wanted to use the helper functions in https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Tosa/Transforms/TosaMakeBroadcastable.cpp, but those are static.

One thing I don't quite understand: Does this mean that the actual input to the `tosa-make-broadcastable` pass invalid TOSA IR? And if the verifier catches rank mismatches, that pass should never change the IR, correct?

Reply:
That's a good point. I think there are two possible choices.

1. Allow invalid broadcast configurations be created. Then rely on tosa-make-broadcastable to convert them to 'TOSA legal' broadcasts by inserting the required tosa.reshape. Finally, add to the 'tosa-validate' pass a check that looks at the shapes for the elementwise ops and makes sure that ranks match.
2. Add a validator to the elementwise ops that checks rank mismatches. This would prevent the ops from existing. It could break some existing legalizations that rely on tosa-make-broadcastable, but has the advantage that created ops are always correct. We could move the make-broadcastable helpers into the proposed construction library from the recent RFC to make legalization easier and deprecate the pass.

We've been going with option 1, without any validation so far. I think 2 has some nice advantages of not having to worry about rank mismatches. It puts slightly more burden on the different legalization passes, but it doesn't feel like a huge burden. So I'm asking legalization changes to start including the rank update to make a transition to 2 easier.

Reply:
FYI, our shape inference for element-wise op has logic to detect broadcast. See `  bool hasNoBroadcast(DimAnalysis *dimAnalysis = nullptr);` in ShapeHelper.hpp. 

We started using dimAnalysis (see its usage in `Conversion/ONNXToKrnl`) to also determine that some runtime dimensions are known to be identical (i.e. relating two `?` at the same rank position in 2 distinct operands).

You could add in the shape helper a function that defines only statically known broadcast, that would probably be helpful in other places... The same logic as in `hasNoBroadcast` could be used to detect that distinct condition.

Just trying to promote common infrastructure so that we all gain from each other's work

Reply:
@AlexandreEichenberger Sorry for the late answer. This sounds like a good suggestion. I would like to address this in a follow up PR. Would that be ok for you?